internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pkgtree
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	pathSeparator = string(filepath.Separator)
+
+	// when walking vendor root hierarchy, ignore file system nodes of the
+	// following types.
+	skipSpecialNodes = os.ModeDevice | os.ModeNamedPipe | os.ModeSocket | os.ModeCharDevice
+)
+
+// DigestFromPathname returns a deterministic hash of the specified file system
+// node, performing a breadth-first traversal of directories. While the
+// specified prefix is joined with the pathname to walk the file system, the
+// prefix string is eliminated from the pathname of the nodes encounted when
+// hashing the pathnames, so that the resultant hash is agnostic to the absolute
+// root directory path of the nodes being checked.
+//
+// This function ignores any file system node named `vendor`, `.bzr`, `.git`,
+// `.hg`, and `.svn`, as these are typically used as Version Control System
+// (VCS) directories.
+//
+// Other than the `vendor` and VCS directories mentioned above, the calculated
+// hash includes the pathname to every discovered file system node, whether it
+// is an empty directory, a non-empty directory, empty file, non-empty file, or
+// symbolic link. If a symbolic link, the referent name is included. If a
+// non-empty file, the file's contents are incuded. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
+// skips symbolic links, and for now, we want the hash to include the symbolic
+// link referents.
+func DigestFromPathname(prefix, pathname string) ([]byte, error) {
```

Comment:
The prefix in this code is merely whatever portion of each pathname must be stripped off in order to not send it to the hash algorithm.

Also keep in mind, as you reminded me, that there is no standard number of directories down from the vendor root to the project location.

For example, when `VerifyDepTree` calls `DigestFromPathname`, it would call `DigestFromPathname("../some/path/to/vendor", "github.com/alice/alice1")`. If we merge those two strings together into `"../some/path/to/vendor/github.com/alice/alice1"`, `DigestFromPathname` would not know which portion of that pathname gets thrown into the hash. We cannot even rely on it to search for `vendor` component, because some environments might actually have `vendor` somewhere on the left of a crazy path.

Even if we have a magic wand that tells us where the project directories are, we do not know how many of the pathname components will be needed for the hash. As humans we know we throw in "github.com/alice/alice1/alice1.go`. I'm not sure how else to send that signal to `DigestFromPathname`.